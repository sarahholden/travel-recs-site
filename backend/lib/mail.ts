import 'dotenv/config';
import {
  createTransport,
  getTestMessageUrl,
  SentMessageInfo,
} from 'nodemailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: 587,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeANiceEmail(text: string) {
  return `
    <div style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>¬©Ô∏è, Sarah Holden</p>
    </div>
  `;
}

type MailResponse = {
  accepted?: string[] | null;
  rejected?: string[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: {
    from: string;
    to?: string[] | null;
  };
  messageId: string;
};

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transport.sendMail({
    to,
    from: 'test@example.com',
    subject: 'Your Sick Fits password reset token!',
    html: makeANiceEmail(`Your Password reset token is here!
      <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click here to reset</a>
    `),
  })) as MailResponse;

  if (process.env.MAIL_USER.includes('ethereal.email')) {
    /* eslint no-console: 0 */
    console.log(
      'üì¨ Message Sent! Preview it at',
      getTestMessageUrl(info as SentMessageInfo)
    );
  }
}
