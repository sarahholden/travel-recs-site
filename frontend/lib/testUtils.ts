import casual from "casual";
// import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeGuide = () => ({
  // __typename: 'Item',
  id: "abc123",
  user: null,
  altText: "New York City Skyline",
  destinations: [],
  image: {
    publicUrlTransformed: "dog.jpg",
  },
  location_info: "The biggest apple in existance",
  name: "New York City",
  status: "AVAILABLE",
});

const fakeUser = (overrides) => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2022-12-11T20:16:13.797Z",
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  product: fakeGuide(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: "aggregate",
            aggregate: {
              count: length,
              __typename: "count",
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeGuide,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
