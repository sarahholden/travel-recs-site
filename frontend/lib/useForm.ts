import { useState } from "react"

type GuideInputsType = {
  name?: string;
  location_info?: string;
}

type ElementPropsType = {
  name: string;
  type: string;
  value: string | number | File
}

export default function useForm(initial: GuideInputsType = {}) {
  const [inputs, setInputs] = useState(initial)

  
  function handleChange(e: React.ChangeEvent<HTMLInputElement> & React.ChangeEvent<HTMLTextAreaElement>) {
    let {name, type, value}: ElementPropsType = e.target;

    if (type === 'number') {
      value = parseInt(value)
    }

    if (type === 'file' && e?.target.files) {
      value = e.target.files[0]
    }

    setInputs({
      ...inputs,
      [name]: value
    })
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(Object.entries(inputs).map(([key, value]) => [key, '']))
    setInputs(blankState)
  }

  return { inputs, handleChange, resetForm, clearForm }
  
}