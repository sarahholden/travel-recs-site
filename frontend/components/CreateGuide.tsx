import useForm from "../lib/useForm";
import Router from "next/router";
import { Form } from "./styles/FormStyles";
import {
  useCreateGuideMutation,
  refetchAllGuidesQueryQuery,
} from "../types/generated-queries";
import DisplayError from "./ErrorMessage";
import useUser from "../hooks/User";
import SignIn from "./SignIn";

export default function CreateGuide() {
  const user = useUser();
  const { inputs, handleChange, resetForm, clearForm } = useForm();

  if (!user) {
    return <SignIn />;
  }

  const [createGuideMutation, { data, error, loading }] =
    useCreateGuideMutation({
      variables: {
        name: inputs.name,
        location_info: inputs.location_info,
        image: inputs.image,
        altText: inputs.altText,
      },
      refetchQueries: [refetchAllGuidesQueryQuery()],
    });

  async function handleSubmit(event: React.FormEvent) {
    event.preventDefault();
    const { data } = await createGuideMutation();
    clearForm();

    data?.createGuide?.id &&
      void Router.push({
        pathname: `/guide/${data.createGuide.id}`,
      });
  }

  return (
    <div>
      <Form action="" onSubmit={handleSubmit}>
        {error && <DisplayError error={error} />}
        <fieldset disabled={loading} aria-busy={loading}>
          <div>
            <label htmlFor="name">Name</label>
            <input
              type="text"
              id="name"
              name="name"
              placeholder="City Name"
              value={inputs.name}
              required
              onChange={handleChange}
            />
          </div>
          <div>
            <label htmlFor="location_info">Location Info</label>
            <textarea
              id="location_info"
              name="location_info"
              placeholder="Location Info"
              value={inputs.location_info}
              onChange={handleChange}
            />
          </div>
          <div>
            <label htmlFor="image">Location Image</label>
            <input
              type="file"
              id="image"
              name="image"
              onChange={handleChange}
            />
          </div>
          <div>
            <label htmlFor="altText">Image Alt Text</label>
            <input
              type="text"
              id="altText"
              name="altText"
              placeholder="Alt Text"
              value={inputs.altText}
              onChange={handleChange}
            />
          </div>

          <div>
            <button type="button" onClick={clearForm}>
              Clear Form
            </button>
          </div>
          <div>
            <button type="submit">Add Guide</button>
          </div>
        </fieldset>
      </Form>
    </div>
  );
}
