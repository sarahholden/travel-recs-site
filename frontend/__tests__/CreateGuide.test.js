import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import userEvent from "@testing-library/user-event";
import Router from 'next/router'; 
import CreateGuide from '../components/CreateGuide'
import {
  CreateGuideDocument,
  AllGuidesQueryDocument,
} from "../types/generated-queries";
import { fakeGuide } from "../lib/testUtils";

const item = fakeGuide();


describe('<CreateGuide/>', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <CreateGuide />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('handles the updating', async () => {
    // 1. Render the form out
    const { container, debug } = render(
      <MockedProvider>
        <CreateGuide />
      </MockedProvider>
    );
    // 2. Type into the boxes
    debug()
    await userEvent.type(
      screen.getByRole("textbox", {
        name: /name/i,
      }),
      item.name
    );
    // 3. Check that those boxes are populated!
  })
})