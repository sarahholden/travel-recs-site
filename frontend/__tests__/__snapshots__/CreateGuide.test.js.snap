// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateGuide/> renders and matches snapshot 1`] = `
<div>
  <h2>
    Sign In
  </h2>
  <form
    class="FormStyles__Form-sc-sf5fob-0 KLHJ"
    method="POST"
  >
    <fieldset
      aria-busy="false"
    >
      <div>
        <label
          for="email"
        >
          Email Address
        </label>
        <input
          id="email"
          name="email"
          placeholder="Enter Your Email"
          type="email"
          value=""
        />
      </div>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          id="password"
          name="password"
          placeholder="Enter Your password"
          type="password"
          value=""
        />
      </div>
      <div>
        <button
          type="submit"
        >
          Sign In
        </button>
      </div>
      <div>
        <p>
          <a
            href="/signup"
          >
            Sign Up
          </a>
        </p>
        <p>
          <a
            href="/requestreset"
          >
            Reset Password
          </a>
        </p>
      </div>
    </fieldset>
  </form>
</div>
`;
